### DISCUSSION TITLE
Set icon in sections/external links in the navigation

### DISCUSSION THREAD

**Discussion by joapuiib** [squidfunk/mkdocs-material#7514] [Q&A] [answered]
I would like to set an icon in the section titles in the navigation, just as icons set in pages via `page.meta.icon`.

![image](https://github.com/user-attachments/assets/8c6b1512-5ba5-44a5-b4fc-af688b17a4ed)

> In the image, I would like to set an icon before DAM1s and DAW1

I have looked at the code, but it seems it's only supported in pages.
https://github.com/squidfunk/mkdocs-material/blob/8df393d11158ebc87026e4eea9ed0c887eae2371/material/templates/partials/nav-item.html#L14-L16

I could try to override that macro to include icons in sections, but my question is: __Which would be the best way to specify the icon in a section?__

I have thought of different options:
- Specify it in the `config.nav`: That would imply changes in the nav config option in Mkdocs, so I'm not sure if it's the best option.
- Use the markdown `:icon:` syntax. That would require some kind of parsing. Could it be done in the template?
- Use a `.meta` file in the section directory. This would require some plugin or hook to read its contents and store it somewhere. The [Section class](https://www.mkdocs.org/dev-guide/themes/#section) doesn't have a `meta` attribute.

**Comment by kamilkrzyskow** [comment]
Hello @joapuiib,

Sections don't have any metadata by default, and the main issue is **how to store and link the icon to a specific section**, I believe there is only the `title` you could use as the Section metadata to distinguish between them, and this could lead to some sort of "side effect" when 2 sections have the same title. I don't think there is any best option, of adding this connection, as most of them are somewhat intrusive.

If the side effect of duplicate Section titles doesn't concern you, my suggestion would be to make use of either the `theme` or `extra` config options in the `mkdocs.yml` file to store and link icons to Section titles. To mitigate the side effect you could create a direct (mirror) copy of the navigation tree inside `theme` or `extra` which would be a second layer of connection, because you could use the `parent` node and their title to distinguish between Sections.

---

> Specify it in the config.nav: That would imply changes in the nav config option in Mkdocs, so I'm not sure if it's the best option.

IIRC you're making your own theme, extending the Material theme, so you could go the route of modifying the MkDocs `nav` logic to support an `icon` value and read it inside the template, but this would involve overriding the `nav` logic before it's loaded from the `mkdocs.yml`, so somebody could see it as too hacky. You could also try extending the logic of [awesome-pages-plugin](https://github.com/lukasgeiter/mkdocs-awesome-pages-plugin) to both support such value in the `.pages` files and include it in the final Navigation object.

> Use a .meta file in the section directory. This would require some plugin or hook to read its contents and store it somewhere. The [Section class](https://www.mkdocs.org/dev-guide/themes/#section) doesn't have a meta attribute.

This would be the less hacky approach, as you would work on the parsed and validated Navigation object in the event loop and would simply add the data from the `.meta` file. I think there is no connection with the directory structure and the navigation tree itself, so I would also recommend to create a duplicate mirror of the navigation tree in the `.meta` file.

> Use the markdown :icon: syntax. That would require some kind of parsing. Could it be done in the template?

I don't think this approach is needed, as the `include` directive allows to add the SVG files? 
The processing of Markdown syntax is done after the `on_page_markdown` event, and it's not available for the templates by default. You could likely add a `filter` function, similar to how there is the `date` filter for the Blog plugin, to pass in such text value and parse it, but this doesn't solve the issue where should the value be read from.
- https://github.com/squidfunk/mkdocs-material/blob/8df393d11158ebc87026e4eea9ed0c887eae2371/src/plugins/blog/plugin.py#L323
- https://github.com/squidfunk/mkdocs-material/blob/8df393d11158ebc87026e4eea9ed0c887eae2371/src/templates/blog-post.html#L96

**Reply by joapuiib** [reply, marked as answer]
Thanks for your detailed answer! It is really inspiring!
I've managed to do it with `awesome-pages-plugin` and the `filter` idea.

- First of all, I have written a plugin that creates two filters `extract_icon` and `remove_icon`:
```python
from mkdocs.plugins import BasePlugin
import re

# This regex allows to match a icon with no text
ICON_RE = re.compile(r'^:(\S+):(?: ([^\n]+))?$')


def extract_icon(section_title):
    match = ICON_RE.match(section_title)
    if match:
        return match.group(1)
    return None


def remove_icon(section_title):
    match = ICON_RE.match(section_title)
    if match:
        return match.group(2) or ''
    return section_title


class SectionIconsPlugin(BasePlugin):
    def on_env(self, env, config, files):
        if not self.config.get('enabled', True):
            return

        env.filters['extract_icon'] = extract_icon
        env.filters['remove_icon'] = remove_icon
```

This plugin parses icons at the start of the title with the syntax `:icon: Section title`.

- With those filers, I've modified the `render_content` macro in the `nav_item.html` partial, so the icon can be extracted from the section title.
```html
{% macro render_content(nav_item, ref = nav_item) %}
  {% set title = ref.title %}
  {% if nav_item.is_page and nav_item.meta.icon %}
    {% include ".icons/" ~ nav_item.meta.icon ~ ".svg" %}
  {% endif %}
  {% if "sectionicons" in plugins and nav_item.is_section %}
    {% set section_icon = ref.title | extract_icon %}
    {% if section_icon %}
      {% set title = ref.title | remove_icon %}
      {% include ".icons/" ~ section_icon ~ ".svg" %}
    {% endif %}
  {% endif %}
  <span class="md-ellipsis">
    {{ title }}
  </span>
  {% if nav_item.is_page and nav_item.meta.status %}
    {{ render_status(nav_item, nav_item.meta.status) }}
  {% endif %}
{% endmacro %}
```
- I've defined the title in the `.pages` files using `awesome-pages-plugin`.
```yaml
nav:
    - ...
    - ":material/pencil-outline: Section title": section_directory
```
> I think it should work without `awesome-pages-plugin` if the section title is specified in `config`, but I've not tested it.

- It works! :smile:

![image](https://github.com/user-attachments/assets/e8f53fd9-d359-4153-b5dc-880622677ac0)

I could write in a separate plugin package, but in order to work, people should have to modify the `nav_item.html` partial.

Gently pinging @squidfunk. Do you think this feature could be useful?

> Any considerations/fixes about the syntax or implementation made are more than welcome.

---

While working on this, I got a new doubt.

In order to override the `render_content` macro inside the `nav_item.html` partial, I have had to copy the entire original file inside my custom theme, so the other macros are still available, and then, make my modifications.

__Is there a way to override a single macro and still use the previous defined macros?__.

**Reply by kamilkrzyskow** [reply]
Very nice solution 👍, once again less complicated than mine 😅. Thanks for sharing ❤️ 

As the title contains a special token that gets stripped afterwards it should also work without the awesome-pages-plugin.
Also it could be simplified with a single `:i:` / `;` separator to not require regex extraction.

> Is there a way to override a single macro and still use the previous defined macros?

I don't think so, no, but then again I never tried it 🤔 

The Jinja2 templates are parsed into a Python script blob, and the defined blocks in the `base.html` can be overridden in the files that `extend` it. Also when developing [mkdocs-nype](https://github.com/nypesap/mkdocs-nype) theme I tested the double extend chain of Material `base.html` -> Nype `nype-base.html` -> Project override `main.html` and the block overrides are still working, so the block scopes are properly being managed at a global level when working from top to bottom of the import chain. 

However, I don't know how are the macros, that are defined in a lower level "node", defined in the final parsed code.

**Reply by joapuiib** [reply]
Thanks for the insight!

I'll try to investigate and I'll let you know if I figure something out.

Many thanks for your help ❤️

**Reply by joapuiib** [reply]
Just in case it helps someone, I've modified the template so it allows icons in external links in the navigation:

```md
{% macro render_content(nav_item, ref = nav_item) %}
  {% set title = ref.title %}
  {% if nav_item.is_page and nav_item.meta.icon %}
    {% include ".icons/" ~ nav_item.meta.icon ~ ".svg" %}
  {% endif %}
  {% if "sectionicons" in config.plugins and (nav_item.is_section or nav_item.is_link) %}
    {% set section_icon = ref.title | extract_icon %}
    {% if section_icon %}
      {% set title = ref.title | remove_icon %}
      {% include ".icons/" ~ section_icon ~ ".svg" %}
    {% endif %}
  {% endif %}
  <span class="md-ellipsis">
    {{ title }}
  </span>
  {% if nav_item.is_page and nav_item.meta.status %}
    {{ render_status(nav_item, nav_item.meta.status) }}
  {% endif %}
{% endmacro %}
```

No change in the plugin is needed.

**Reply by joapuiib** [reply]
In order to set the icon in the section title on mobile screens, an adjustment has been made:

![image](https://github.com/user-attachments/assets/a1137c4a-49d7-481e-a0d3-1a473acf2927)

In __nav-item.html__ partial:
https://github.com/squidfunk/mkdocs-material/blob/47fc7d5a61847b9a184facb1cea3e356e0bdcee5/src/templates/partials/nav-item.html#L214-L216
```html
<nav class="md-nav" data-md-level="{{ level }}" aria-labelledby="{{ path }}_label" aria-expanded="{{ nav_item.active | tojson }}">
  <label class="md-nav__title" for="{{ path }}">
    <span class="md-nav__icon md-icon"></span>
    {% set title = nav_item.title %}
    {% if "sectionicons" in config.plugins and nav_item.is_section %}
      {% set section_icon = nav_item.title | extract_icon %}
      {% if section_icon %}
        {% set title = nav_item.title | remove_icon %}
        {% include ".icons/" ~ section_icon ~ ".svg" %}
      {% endif %}
    {% endif %}
    {{ title }}
  </label>
```

In this case, a CSS is needed to set the size of the icon:

```css
.md-nav__title svg {
    fill: currentcolor;
    flex-shrink: 0;
    height: 1.3em;
    position: relative;
}
```
This styles are a copy of:
https://github.com/squidfunk/mkdocs-material/blob/47fc7d5a61847b9a184facb1cea3e356e0bdcee5/src/templates/assets/stylesheets/main/components/_nav.scss#L120-L128

---

Just in case, the full __nav-item.html__ partial with the modifications can be found at [my custom theme](https://github.com/joapuiib/mkdocs-material-joapuiib/blob/main/material_joapuiib/templates/partials/nav-item.html).

**Comment by ojacques** [comment]
Great discussion and insights. Another way to achieve this is through awesome-pages plugin, and using Unicode emojis in the alternate title, which are now largely supported by systems (client side). On Windows, you get them with the windows key + ";". The challenge is that they are not coming from the material icon library.

Example of a `.pages`:

```yaml
nav:
  - "🏠 DOFLinx": README.md
  - "🚀 Getting Started": getting-started
  - "⚙️ Parameters": parameters
  - "🕹️ Systems": systems
  - "🌟 Misc": Misc
  - "🆘 Help": 07_Help.md
  - ...
```

![image](https://github.com/user-attachments/assets/9088117b-d841-4279-91ff-0bd9533a53f1)
